File: services/api/app/schemas/dataset_validation.py
"""Response schemas (DTOs) for the dataset validation API.

The engine's own dataclasses (``ValidationReport``, ``ValidationIssue``, …)
are deliberately Pydantic-free so rules stay independent of the web layer;
this module is the one place those are mapped to the wire — the same split
``app/schemas/features.py`` and ``app/schemas/indicators.py`` already make.

``DatasetValidationRequest`` is a ``FeatureDatasetRequest`` (imported, not
redeclared) plus two validation-only fields. Subclassing rather than
duplicating the market/timeframe/range/feature-selection fields is what
lets the exact same request body that builds a dataset also validate it —
a caller changes only the URL, not the shape of what they send.
"""

from datetime import UTC, datetime
from typing import Any

from pydantic import BaseModel, Field, field_serializer

from app.dataset_validation.base import ValidationIssue, ValidationRuleMetadata
from app.dataset_validation.report import CategorySummary, ValidationReport, ValidationSummary
from app.schemas.features import FeatureDatasetRequest


class DatasetValidationRequest(FeatureDatasetRequest):
    """Everything needed to build a dataset, plus what to validate it against."""

    required_columns: list[str] = Field(
        default_factory=list,
        description="Column names that must be present, beyond what the features imply",
    )
    rules: list[str] | None = Field(
        default=None,
        description="Run only these rule names; omit to run every registered rule",
    )


class ValidationIssueDTO(BaseModel):
    """One finding from one rule."""

    rule: str
    category: str
    severity: str
    code: str
    message: str
    column: str | None = None
    row_index: int | None = None
    count: int | None = None
    details: dict[str, Any] = Field(default_factory=dict)

    @classmethod
    def from_issue(cls, issue: ValidationIssue) -> "ValidationIssueDTO":
        """Map an engine ``ValidationIssue`` onto the wire DTO."""
        return cls(
            rule=issue.rule,
            category=issue.category,
            severity=issue.severity,
            code=issue.code,
            message=issue.message,
            column=issue.column,
            row_index=issue.row_index,
            count=issue.count,
            details=issue.details,
        )


class ValidationSummaryDTO(BaseModel):
    """Headline counts across every issue found, regardless of category."""

    total_checks: int
    errors: int
    warnings: int
    info: int

    @classmethod
    def from_summary(cls, summary: ValidationSummary) -> "ValidationSummaryDTO":
        """Map an engine ``ValidationSummary`` onto the wire DTO."""
        return cls(
            total_checks=summary.total_checks,
            errors=summary.errors,
            warnings=summary.warnings,
            info=summary.info,
        )


class CategorySummaryDTO(BaseModel):
    """Issue counts scoped to one validation category."""

    errors: int
    warnings: int
    info: int

    @classmethod
    def from_summary(cls, summary: CategorySummary) -> "CategorySummaryDTO":
        """Map an engine ``CategorySummary`` onto the wire DTO."""
        return cls(errors=summary.errors, warnings=summary.warnings, info=summary.info)


class ValidationReportResponse(BaseModel):
    """The outcome of validating one dataset: pass/fail, every issue, and how long it took."""

    dataset_id: str = Field(..., description="The validated dataset's own unique build id")
    symbol: str
    timeframe: str
    engine_version: str = Field(
        ..., description="Version of the validation engine itself, independent of any rule's own"
    )
    validated_at: datetime
    passed: bool = Field(
        ..., description="False only when at least one error-severity issue was found"
    )
    rules_run: list[str] = Field(..., description="Rule names actually run, in run order")
    summary: ValidationSummaryDTO
    categories: dict[str, CategorySummaryDTO] = Field(
        ..., description="Per-category issue counts; every category is present, even at zero"
    )
    issues: list[ValidationIssueDTO]
    rows: int = Field(..., description="Rows in the validated dataset")
    columns: int = Field(..., description="Columns in the validated dataset")
    duration_ms: float = Field(..., description="Wall-clock time spent running every rule")

    @field_serializer("validated_at")
    def _serialize_validated_at(self, value: datetime) -> str:
        """ISO-8601 UTC with a literal ``Z``, matching every timestamp on this API."""
        return value.astimezone(UTC).isoformat().replace("+00:00", "Z")

    @classmethod
    def from_report(cls, report: ValidationReport) -> "ValidationReportResponse":
        """Map an engine ``ValidationReport`` onto the wire DTO."""
        return cls(
            dataset_id=report.dataset_id,
            symbol=report.symbol,
            timeframe=report.timeframe,
            engine_version=report.engine_version,
            validated_at=report.validated_at,
            passed=report.passed,
            rules_run=list(report.rules_run),
            summary=ValidationSummaryDTO.from_summary(report.summary),
            categories={
                name: CategorySummaryDTO.from_summary(summary)
                for name, summary in report.categories.items()
            },
            issues=[ValidationIssueDTO.from_issue(issue) for issue in report.issues],
            rows=report.rows,
            columns=report.columns,
            duration_ms=report.duration_ms,
        )


class ValidationRuleDTO(BaseModel):
    """One validation rule's catalogue entry."""

    name: str
    category: str
    description: str
    default_severity: str
    version: str

    @classmethod
    def from_metadata(cls, metadata: ValidationRuleMetadata) -> "ValidationRuleDTO":
        """Map engine metadata onto the wire DTO."""
        return cls(
            name=metadata.name,
            category=metadata.category,
            description=metadata.description,
            default_severity=metadata.default_severity,
            version=metadata.version,
        )


class ValidationRuleCatalogResponse(BaseModel):
    """Every validation rule the engine can run."""

    rules: list[ValidationRuleDTO]
    total: int = Field(..., description="Number of registered rules")
    categories: list[str] = Field(
        ..., description="Distinct categories present, for grouping in a UI"
    )

File: apps/dashboard/src/types/api/dataset-validation.ts
import { z } from 'zod';

/**
 * Wire schemas for the dataset validation API (see
 * `services/api/app/api/v1/endpoints/dataset_validation.py`). Every
 * response is validated against these before the app trusts it, matching
 * the same convention as `features.ts` and `indicators.ts`.
 */

export const ValidationIssueSchema = z.object({
  rule: z.string(),
  category: z.string(),
  severity: z.enum(['error', 'warning', 'info']),
  code: z.string(),
  message: z.string(),
  column: z.string().nullable().optional(),
  row_index: z.number().int().nullable().optional(),
  count: z.number().int().nullable().optional(),
  details: z.record(z.string(), z.unknown()).optional(),
});

export type ValidationIssue = z.infer<typeof ValidationIssueSchema>;

export const ValidationSummarySchema = z.object({
  total_checks: z.number().int().nonnegative(),
  errors: z.number().int().nonnegative(),
  warnings: z.number().int().nonnegative(),
  info: z.number().int().nonnegative(),
});

export type ValidationSummary = z.infer<typeof ValidationSummarySchema>;

export const CategorySummarySchema = z.object({
  errors: z.number().int().nonnegative(),
  warnings: z.number().int().nonnegative(),
  info: z.number().int().nonnegative(),
});

export type CategorySummary = z.infer<typeof CategorySummarySchema>;

/**
 * The outcome of validating one dataset. `passed` is false only when at
 * least one `error`-severity issue was found — a `warning`/`info` issue is
 * always reported but never fails the gate.
 */
export const ValidationReportSchema = z.object({
  dataset_id: z.string(),
  symbol: z.string(),
  timeframe: z.string(),
  engine_version: z.string(),
  validated_at: z.string().datetime(),
  passed: z.boolean(),
  rules_run: z.array(z.string()),
  summary: ValidationSummarySchema,
  categories: z.record(z.string(), CategorySummarySchema),
  issues: z.array(ValidationIssueSchema),
  rows: z.number().int().nonnegative(),
  columns: z.number().int().nonnegative(),
  duration_ms: z.number(),
});

export type ValidationReport = z.infer<typeof ValidationReportSchema>;

export const ValidationRuleSchema = z.object({
  name: z.string(),
  category: z.string(),
  description: z.string(),
  default_severity: z.enum(['error', 'warning', 'info']),
  version: z.string(),
});

export type ValidationRule = z.infer<typeof ValidationRuleSchema>;

export const ValidationRuleCatalogSchema = z.object({
  rules: z.array(ValidationRuleSchema),
  total: z.number().int().nonnegative(),
  categories: z.array(z.string()),
});

export type ValidationRuleCatalog = z.infer<typeof ValidationRuleCatalogSchema>;
